internal/pkg/gateway/api.go

```diff
@@ -427,47 +427,94 @@ func (gs *Server) Submit(ctx context.Context, request *gp.SubmitRequest) (*gp.Su
 		return nil, status.Errorf(codes.Unavailable, "no orderer nodes available")
 	}
 
-	// try each orderer in random order
 	var errDetails []proto.Message
-	for _, index := range rand.Perm(len(orderers)) {
-		orderer := orderers[index]
-		logger.Infow("Sending transaction to orderer", "txID", request.TransactionId, "endpoint", orderer.logAddress)
-
-		var response *ab.BroadcastResponse
-		var err error
-		done := make(chan struct{})
-		go func() {
-			defer close(done)
-			ctx, cancel := context.WithTimeout(ctx, gs.options.BroadcastTimeout)
-			defer cancel()
-
-			response, err = gs.broadcast(ctx, orderer, txn)
-		}()
-		select {
-		case <-done:
-			// Broadcast completed normally
-		case <-ctx.Done():
-			// Overall submit timeout expired
-			logger.Warn("Submit call timed out while broadcasting to ordering service")
-			return nil, newRpcError(codes.DeadlineExceeded, "submit timeout expired while broadcasting to ordering service")
+	if gs.getChannelConfig(request.ChannelId).ChannelConfig().Capabilities().ConsensusTypeBFT() {
+		// For BFT, need to send transaction to ALL orderers
+		waitCh := make(chan *gp.ErrorDetail, len(orderers))
+		for _, o := range orderers {
+			go func(ord *orderer) {
+				logger.Infow("Sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress)
+				ctx, cancel := context.WithTimeout(ctx, gs.options.BroadcastTimeout)
+				defer cancel()
+				response, err := gs.broadcast(ctx, ord, txn)
+
+				if err != nil {
+					logger.Warnw("Error sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress, "err", err)
+					waitCh <- errorDetail(ord.endpointConfig, err.Error())
+				} else if status := response.GetStatus(); status == common.Status_SUCCESS {
+					waitCh <- nil
+				} else {
+					logger.Warnw("Unsuccessful response sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress, "status", status, "info", response.GetInfo())
+					waitCh <- errorDetail(ord.endpointConfig, fmt.Sprintf("received unsuccessful response from orderer: %s", common.Status_name[int32(status)]))
+				}
+			}(o)
 		}
-
-		if err != nil {
-			errDetails = append(errDetails, errorDetail(orderer.endpointConfig, err.Error()))
-			logger.Warnw("Error sending transaction to orderer", "txID", request.TransactionId, "endpoint", orderer.logAddress, "err", err)
-			continue
+		atLeastOneSuccess := false
+	loop:
+		for i := 0; i < len(orderers); i++ {
+			select {
+			case osnErr := <-waitCh:
+				// Broadcast completed normally
+				if osnErr != nil {
+					errDetails = append(errDetails, osnErr)
+				} else {
+					atLeastOneSuccess = true
```

Comment:
at least one is not enough, we should have at least `2f+1`.

Reply:
@yacovm there could be a variant like this:
- received less than 2f+1 positive responses (e.g., 1), sent an error to the client, but the orderer cluster executed the transaction and added it to the blockchain

Reply:
What you describe is what will happen with our without our control as a gateway. Once a transaction has been sent to an orderer, there is no way to stop it from being included in a block. 

That's why I said that you need to have at least 2f+1 positive responses. 

In practice, if you cannot get 2f+1 positive responses then probably something is wrong with the orderer cluster.

Reply:
Not quite so.
It could be a failure between the gateway and several orderer nodes.
But at the same time there is complete consistency between the orderers.

Reply:
If we return an error, the gateway will not wait for the block to be released and will return an error to the client.

If we return "no error" then the gateway will wait for the block to be released. The block may be released or it may end with a timeout error. 

The timeout error is the worst error. What happened after the timeout error is a separate epic

Reply:
Perhaps an integration test is needed to decide how to proceed?

Reply:
I'll add integration tests once the BFT orderer PR has been merged